### PR TITLE
feat: add replay button

### PR DESCRIPTION
**Pull request by VitorPietrobom**
## Description

add functional replay button and loop button

## Related Issues

- Closes [#676](https://github.com/profusion/quickstart-ios-graphql/issues/676)

## Progress

- [x] Be able to replay a video
- [x] Be able to replay an audio
- [x] Forward controls must be disabled at the end of the video
- [x] (Nice to have) Auto replay (loop)

### Replay
 - Get to the end of a video
 - Press replay

### Loop
- Activate loop
- Get to the end of a video

## Visual reference

https://github.com/profusion/ios-enhanced-media-player/assets/53837251/5faa9489-21c5-42fe-bb9c-4ca9019a4183

### CODE REVIEW

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 1]
```diff
@@ -37,23 +37,41 @@ struct MediaPlayerView: View {
...
+                if viewModel.playerState == .finished {
+                    renderReplay()
+                } else {
+                    renderPlayPause()
+                }
+                renderForward(disabled: (viewModel.playerState == .finished))
+                    .frame(maxWidth: .infinity, alignment: .trailing)
+            }
+            HStack {
             }
```
Comment (pedromotita) [inline EnhancedMediaPlayer/MediaPlayerView.swift:56]: Remove empty `HStack`

Comment (VitorPietrobom) [reply EnhancedMediaPlayer/MediaPlayerView.swift:56]: This is actually to keep the buttons aligned, future implementations will place items inside

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 2]
```diff
@@ -148,6 +170,7 @@ extension MediaPlayerView.Control {
     enum Constants {
         static let playIcon = "play.fill"
         static let pauseIcon = "pause.fill"
+        static let loopIcon = "infinity"
```
[the remark this answers is not in the document]
Comment (VitorPietrobom) [reply EnhancedMediaPlayer/MediaPlayerView.swift:173]: We are already using that for the replay button, do you think it makes sense to change both?

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 3]
```diff
@@ -95,6 +112,7 @@ struct MediaPlayerView: View {
 extension MediaPlayerView {
     class ViewModel: ObservableObject {
         @Published var playerState: PlayerState = .loading
+        @Published var inLoop: Bool = false
         var playerViewHandler: MediaPlayerControllerRepresentable?
```
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:116]: A line between `@Published` and non

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 4]
```diff
@@ -37,23 +37,41 @@ struct MediaPlayerView: View {
...
+
+    @ViewBuilder private func renderLoop() -> some View {
+        if viewModel.inLoop {
+            renderButton(control: .loop, color: .blue)
+                .frame(maxWidth: .infinity, alignment: .trailing)
+        } else {
+            renderButton(control: .loop, color: .white)
+                .frame(maxWidth: .infinity, alignment: .trailing)
+        }
+
```
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:73]: unnecessary whiteline

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 5]
```diff
@@ -37,23 +37,41 @@ struct MediaPlayerView: View {
...
     }
+
+    @ViewBuilder private func renderLoop() -> some View {
+        if viewModel.inLoop {
+            renderButton(control: .loop, color: .blue)
+                .frame(maxWidth: .infinity, alignment: .trailing)
+        } else {
+            renderButton(control: .loop, color: .white)
+                .frame(maxWidth: .infinity, alignment: .trailing)
+        }
```
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:72]: `renderButton(control: .loop, color: viewModel.inLoop ? .blue : .white)`

File: EnhancedMediaPlayer/MediaPlayerView.swift (hunk 1 above)
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:56]: Empty HStack

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 6]
```diff
@@ -37,23 +37,41 @@ struct MediaPlayerView: View {
...
+            }
+            HStack {
+                renderRewind()
+                    .frame(maxWidth: .infinity, alignment: .leading)
+                if viewModel.playerState == .finished {
+                    renderReplay()
+                } else {
+                    renderPlayPause()
+                }
+                renderForward(disabled: (viewModel.playerState == .finished))
```
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:52]: unnecessary inner parenthesis

File: EnhancedMediaPlayer/MediaPlayerView.swift [hunk 7]
```diff
@@ -65,19 +83,18 @@ struct MediaPlayerView: View {
...
+    @ViewBuilder private func renderButton(control: Control, disabled: Bool = false, color: Color = .white) -> some View {
         Button(action: { viewModel.onTapAction?(control) }) {
             control.systemImage
-                .resizable()
-                .frame(width: Constants.iconSize.width, height: Constants.iconSize.height)
-                .foregroundColor(.white)
-        }
+                .font(.system(size: Constants.iconSize))
+                .tint(color)
+        }.disabled(disabled)
```
Comment (vitorjmoriya) [inline EnhancedMediaPlayer/MediaPlayerView.swift:97]:
```
}
.disabled(disabled)

File: EnhancedMediaPlayer/MediaPlayerView.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (leoyt61) [reply EnhancedMediaPlayer/MediaPlayerView.swift:173]: hummm, maybe we could use `gobackward` for the replay and `repeat` for the loop, like in youtube. 🤔

[the remark this answers is not in the document]
Comment (leoyt61) [reply EnhancedMediaPlayer/MediaPlayerView.swift:173]: nvm, just saw that we're already using `gobackward` for the rewind action

**Review by andrepapoti — APPROVED** [review]
Beside the nits that vitor pointed out LGTM, just make sure to tackle those

File: EnhancedMediaPlayer/MediaPlayerView.swift (hunk 1 above)
Comment (vitorjmoriya) [reply EnhancedMediaPlayer/MediaPlayerView.swift:56]: Yeah, but you can use and `.padding` adjustment instead

**Pull request merged**